### PR TITLE
Support OpenAI python SDK v1

### LINKED ISSUE #10793: [BUG] Unable to reproduce OpenAI examples
### Issues Policy acknowledgement

- [X] I have read and agree to submit bug reports in accordance with the [issues policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md)

### Where did you encounter this bug?

Local machine

### Willingness to contribute

Yes. I would be willing to contribute a fix for this bug with guidance from the MLflow community.

### MLflow version

- Client: 2.9.2
- Tracking server: 2.9.2

### System information

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: 22.04
- **Python version**: 3.8.10
- **yarn version, if running the dev UI**:

### Describe the problem

I am not able to reproduce the example in OpenAI flavor in order to log a model for an `openai.ChatCompletion` task, as is directed [here](https://mlflow.org/docs/latest/llms/openai/guide/index.html).

Using openai == 1.6.1

### Tracking information

```shell

```

### Code to reproduce issue

```
import mlflow
import openai
from mlflow.models.signature import ModelSignature
from mlflow.types.schema import ColSpec, ParamSchema, ParamSpec, Schema

mlflow.set_experiment("chat")

with mlflow.start_run():
    model_info = mlflow.openai.log_model(
        model="gpt-3.5-turbo",
        task=openai.ChatCompletion,
        artifact_path="model",
        messages=[{"role": "user", "content": "Tell me a joke about {animal}."}],
        signature=ModelSignature(
            inputs=Schema([ColSpec(type="string", name=None)]),
            outputs=Schema([ColSpec(type="string", name=None)]),
            params=ParamSchema(
                [
                    ParamSpec(name="temperature", default=0, dtype="float"),
                ]
            ),
        ),
    )
```

### Stack trace

```
---------------------------------------------------------------------------
MlflowException                           Traceback (most recent call last)
Cell In [6], line 2
      1 with mlflow.start_run():
----> 2     model_info = mlflow.openai.log_model(
      3         model="gpt-3.5-turbo",
      4         task=openai.ChatCompletion,
      5         artifact_path="model",
      6         messages=[{"role": "user", "content": "Tell me a joke about {animal}."}],
      7         signature=ModelSignature(
      8             inputs=Schema([ColSpec(type="string", name=None)]),
      9             outputs=Schema([ColSpec(type="string", name=None)]),
     10             params=ParamSchema(
     11                 [
     12                     ParamSpec(name="temperature", default=0, dtype="float"),
     13                 ]
     14             ),
     15         ),
     16     )

File ~/.cache/pypoetry/virtualenvs/banner-categorization-flTayMDf-py3.8/lib/python3.8/site-packages/mlflow/openai/__init__.py:540, in log_model(model, task, artifact_path, conda_env, code_paths, registered_model_name, signature, input_example, await_registration_for, pip_requirements, extra_pip_requirements, metadata, **kwargs)
    448 @experimental
    449 @format_docstring(LOG_MODEL_PARAM_DOCS.format(package_name=FLAVOR_NAME))
    450 def log_model(
   (...)
    463     **kwargs,
    464 ):
    465     """
    466     Log an OpenAI model as an MLflow artifact for the current run.
    467 
   (...)
    538 
    539     """
--> 540     return Model.log(
    541         artifact_path=artifact_path,
    542         flavor=mlflow.openai,
    543         registered_model_name=registered_model_name,
    544         model=model,
    545         task=task,
    546         conda_env=conda_env,
    547         code_paths=code_paths,
    548         signature=signature,
    549         input_example=input_example,
    550         await_registration_for=await_registration_for,
    551         pip_requirements=pip_requirements,
    552         extra_pip_requirements=extra_pip_requirements,
    553         metadata=metadata,
    554         **kwargs,
    555     )

File ~/.cache/pypoetry/virtualenvs/banner-categorization-flTayMDf-py3.8/lib/python3.8/site-packages/mlflow/models/model.py:619, in Model.log(cls, artifact_path, flavor, registered_model_name, await_registration_for, metadata, **kwargs)
    613 if (
    614     (tracking_uri == "databricks" or get_uri_scheme(tracking_uri) == "databricks")
    615     and kwargs.get("signature") is None
    616     and kwargs.get("input_example") is None
    617 ):
    618     _logger.warning(_LOG_MODEL_MISSING_SIGNATURE_WARNING)
--> 619 flavor.save_model(path=local_path, mlflow_model=mlflow_model, **kwargs)
    620 mlflow.tracking.fluent.log_artifacts(local_path, mlflow_model.artifact_path)
    621 try:

File ~/.cache/pypoetry/virtualenvs/banner-categorization-flTayMDf-py3.8/lib/python3.8/site-packages/mlflow/openai/__init__.py:340, in save_model(model, task, path, conda_env, code_paths, mlflow_model, signature, input_example, pip_requirements, extra_pip_requirements, metadata, **kwargs)
    338 _validate_and_prepare_target_save_path(path)
    339 code_dir_subpath = _validate_and_copy_code_paths(code_paths, path)
--> 340 task = _get_task_name(task)
    342 if mlflow_model is None:
    343     mlflow_model = Model()

File ~/.cache/pypoetry/virtualenvs/banner-categorization-flTayMDf-py3.8/lib/python3.8/site-packages/mlflow/openai/__init__.py:177, in _get_task_name(task)
    175     return _class_to_task(task)
    176 else:
--> 177     raise mlflow.MlflowException(
    178         f"Unsupported task type: {type(task)}", error_code=INVALID_PARAMETER_VALUE
    179     )

MlflowException: Unsupported task type: <class 'openai.lib._old_api.APIRemovedInV1Proxy'>
```

### Other info / logs

```
/home/sniafas/.cache/pypoetry/virtualenvs/banner-categorization-flTayMDf-py3.8/lib/python3.8/site-packages/pydantic/_internal/_config.py:321: UserWarning: Valid config keys have changed in V2:
* 'schema_extra' has been renamed to 'json_schema_extra'
  warnings.warn(message, UserWarning)
[2024-01-08 18:24:39 +0200] [407218] [INFO] Starting gunicorn 21.2.0
[2024-01-08 18:24:39 +0200] [407218] [INFO] Listening at: http://127.0.0.1:5000 (407218)
[2024-01-08 18:24:39 +0200] [407218] [INFO] Using worker: sync
[2024-01-08 18:24:39 +0200] [407220] [INFO] Booting worker with pid: 407220
[2024-01-08 18:24:39 +0200] [407221] [INFO] Booting worker with pid: 407221
[2024-01-08 18:24:39 +0200] [407222] [INFO] Booting worker with pid: 407222
```

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [X] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [X] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### LINKED ISSUE #10810: [BUG] mlflow.lopenai.log_model tasks unsupported
### Issues Policy acknowledgement

- [X] I have read and agree to submit bug reports in accordance with the [issues policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md)

### Where did you encounter this bug?

Databricks

### Willingness to contribute

Yes. I would be willing to contribute a fix for this bug with guidance from the MLflow community.

### MLflow version

- mlflow 2.9.2
- mlflow-sklinny 2.8.1

### System information

- Standard_DS3_v2
- DBR 14.2 ML

### Describe the problem

Am able to run the same code locally and have it work.  Trying to run a mlflow.openai.log_model in a Databricks notebook and getting error `MlflowException: Unsupported task type: <class 'openai.lib._old_api.APIRemovedInV1Proxy'>
`

Get messages that both openai.ChatCompletions and openai.Completions tasks are unsupported

### Tracking information

```shell
REPLACE_ME
```

### Code to reproduce issue

```
import openai


system_prompt = (
    "You are a helpful assistant."
)
        
mlflow.log_param("system_prompt", system_prompt)

# Evaluate the model on some example questions
questions = pd.DataFrame(
    {
        "questions": [
            "Who won the Super Bowl?"
        ],
    }
)
logged_model = mlflow.openai.log_model(
    model="gpt-3.5-turbo-1106",
    task=openai.ChatCompletion,
    artifact_path="model",
    context = query(questions),
    messages=[
        {"role": "system", "content": system_prompt},
        {"role": "user", "content": f"The user questions is:{questions} \n The context:{query(questions)}"},
    ],
)


mlflow.evaluate(
    model=logged_model.model_uri,
    model_type="question-answering",
    data=questions,
)

# Load and inspect the evaluation results
results: pd.DataFrame = mlflow.load_table(
    "eval_results_table.json", extra_columns=["run_id", "params.system_prompt"]
)
print("Evaluation results:")
print(results)
```

### Stack trace

```
File /databricks/python/lib/python3.10/site-packages/mlflow/openai/__init__.py:571, in log_model(model, task, artifact_path, conda_env, code_paths, registered_model_name, signature, input_example, await_registration_for, pip_requirements, extra_pip_requirements, metadata, **kwargs)
    479 @experimental
    480 @format_docstring(LOG_MODEL_PARAM_DOCS.format(package_name=FLAVOR_NAME))
    481 def log_model(
   (...)
    494     **kwargs,
    495 ):
    496     """
    497     Log an OpenAI model as an MLflow artifact for the current run.
    498 
   (...)
    569 
    570     """
--> 571     return Model.log(
    572         artifact_path=artifact_path,
    573         flavor=mlflow.openai,
    574         registered_model_name=registered_model_name,
    575         model=model,
    576         task=task,
    577         conda_env=conda_env,
    578         code_paths=code_paths,
    579         signature=signature,
    580         input_example=input_example,
    581         await_registration_for=await_registration_for,
    582         pip_requirements=pip_requirements,
    583         extra_pip_requirements=extra_pip_requirements,
    584         metadata=metadata,
    585         **kwargs,
    586     )

File /databricks/python/lib/python3.10/site-packages/mlflow/models/model.py:619, in Model.log(cls, artifact_path, flavor, registered_model_name, await_registration_for, metadata, **kwargs)
    613 if (
    614     (tracking_uri == "databricks" or get_uri_scheme(tracking_uri) == "databricks")
    615     and kwargs.get("signature") is None
    616     and kwargs.get("input_example") is None
    617 ):
    618     _logger.warning(_LOG_MODEL_MISSING_SIGNATURE_WARNING)
--> 619 flavor.save_model(path=local_path, mlflow_model=mlflow_model, **kwargs)
    620 mlflow.tracking.fluent.log_artifacts(local_path, mlflow_model.artifact_path)
    621 try:

File /databricks/python/lib/python3.10/site-packages/mlflow/openai/__init__.py:371, in save_model(model, task, path, conda_env, code_paths, mlflow_model, signature, input_example, pip_requirements, extra_pip_requirements, metadata, **kwargs)
    369 _validate_and_prepare_target_save_path(path)
    370 code_dir_subpath = _validate_and_copy_code_paths(code_paths, path)
--> 371 task = _get_task_name(task)
    373 if mlflow_model is None:
    374     mlflow_model = Model()

File /databricks/python/lib/python3.10/site-packages/mlflow/openai/__init__.py:181, in _get_task_name(task)
    179     return _class_to_task(task)
    180 else:
--> 181     raise mlflow.MlflowException(
    182         f"Unsupported task type: {type(task)}", error_code=INVALID_PARAMETER_VALUE
    183     )

```

### Other info / logs

```
REPLACE_ME
```

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [X] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [X] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by harupy**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/harupy/mlflow/pull/11034?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/11034/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 11034
```

</p>
</details>

### Related Issues/PRs

Resolve #10793
Resolve #10810

### What changes are proposed in this pull request?

Summary

- Legacy OpenAI objects were updated:
  - `openai.Embedding` -> `openai.emebeddings`
  - `openai.ChatCompletions` -> `openai.chat.completions`
  - `openai.Comletions` -> `openai.completions`
- Created a fake OpenAI service. v0 uses `requests` for making HTTP requests while v1 uses `httpx`. Creating a fake OpenAI service is easier and cleaner than monkey-patching both `requests` and `httpx` during tests.
- No updates in the prediction logic.
- Both v1 and v0 are supported.

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/openai/__init__.py [hunk 1]
```diff
@@ -812,39 +811,13 @@ def predict(self, data, params: Optional[Dict[str, Any]] = None):
...
-                .. Note:: Experimental: This parameter may change or be removed in a future
-                           release without warning.
-
-        Returns:
-            Model predictions.
-        """
-        from mlflow.utils.openai_utils import _mock_openai_request
-
-        with _mock_openai_request():
-            return super().predict(data)
```
Comment (harupy) [inline mlflow/openai/__init__.py:837]: We have the fake openai service. This class is no longer needed.

File: mlflow/openai/__init__.py [hunk 2]
```diff
@@ -706,17 +705,17 @@ def get_params_list(self, data):
...
             deployment_id = self.request_configs.get("deployment_id")
 
             return (
                 f"{api_base}/openai/deployments/{deployment_id}/"
                 f"{task_url}?api-version={api_version}"
             )
-        else:
-            return default_url
+
+        return f"{api_base}/{task_url}" if api_base else default_url
```
Comment (harupy) [inline mlflow/openai/__init__.py:718]: A small but important change. In the original code, `api_base` is only used for Azure.

File: mlflow/openai/__init__.py [hunk 3]
```diff
@@ -114,71 +115,70 @@ def get_default_conda_env():
...
+            openai.chat.completions: "chat.completions",
+            openai.completions: "completions",
+            openai.images.edit: "images.edit",
+            openai.embeddings: "embeddings",
+            openai.files: "files",
+            openai.images: "images",
+            openai.fine_tuning: "fine_tuning",
+            openai.moderations: "moderations",
+            openai.models: "models",
+        }
```
Comment (harupy) [inline mlflow/openai/__init__.py:150]: In V1, `openai.Model`, `openai.Engine`, and `openai.Deployment` were removed.

File: mlflow/openai/__init__.py [hunk 4]
```diff
@@ -114,71 +115,70 @@ def get_default_conda_env():
...
     else:
-        raise mlflow.MlflowException(
-            f"Unsupported task type: {type(task)}", error_code=INVALID_PARAMETER_VALUE
-        )
+        task_name = mapping.get(task)
+        if task_name is None:
+            raise mlflow.MlflowException(
+                f"Unsupported task object: {task}", error_code=INVALID_PARAMETER_VALUE
+            )
+    return task_name
```
Comment (serena-ruan) [inline mlflow/openai/__init__.py:181]:
```suggestion
        return task_name
```

**Review by serena-ruan — APPROVED** [review]
LGTM!

**Review by B-Step62 — APPROVED** [review]
LGTM!

**Pull request merged**